torch/csrc/jit/interpreter.cpp

```diff
@@ -283,97 +592,189 @@ struct Instruction {
+          auto then_block_start = instructions.size();
+          insertNodesFromBlock(then_block);
+          insertAssign(then_block->outputs(), moveFlags(then_block), node->outputs());
+          createJump(jump, instructions.size());
+          createJumpNZ(cond_branch, then_block_start);
+        } break;
+        case kLoop: {
+          // o0 = while c i0
+          //        block 0: l0
+          //          <body>
+          //          -> (v0, v1)
+
+          // turns into:
+          // l0 = i0
+          // JumpZ c, end
+          // begin:
+          //   <body>
+          //   c, l0 = v0, v1
+          //   JumpNZ c, begin
+          // end:
+
+          auto body_block = node->blocks()[0];
+
+          // before assign op: stack: ... <cond> <loop-carried-depdencies>
+          insertAssign(node->inputs(), moveFlags(node), body_block->inputs());
+          // after assign op: stack: ... <cond>
+          // cond_branch consumes <cond> from top of the stack
+          auto cond_branch = insertInstruction(kPlaceholder, {}, {}, {});
+          // after branch: stack: ...
+
+          auto entry = instructions.size();
+          insertNodesFromBlock(body_block);
+          // before assign op: stack: ... <cond> <loop-carried-depdencies>
+          insertAssign(body_block->outputs(), moveFlags(body_block), body_block->inputs());
+          // after assign op: stack: ... <cond>
+          auto cond_branch_end = insertInstruction(kPlaceholder, {}, {}, {});
+          // after branch: stack: ...
+
+          aliasRegistersTo(node->outputs(), body_block->inputs());
+          createJumpZ(cond_branch, instructions.size());
+          createJumpNZ(cond_branch_end, entry);
+        } break;
+        default: {
+          insertInstruction(node);
+        } break;
       }
-      listBegin(stage.outputs.values);
-      for(;output_pos < graph->outputs().size(); output_pos++) {
-        auto output = graph->outputs()[output_pos];
-        if((int64_t)output->stage() > cur_stage)
-          break;
-        listInsert(stage.outputs.values, getOrAllocateRegister(output));
+      // each stage ends with a load instruction
+      // we record where these instructions occur, and use them to
+      // exit the interpreter
+      if(node->kind() == kLoad) {
+        stage_end.push_back(instructions.size());
       }
     }
   }
+
+  size_t insertInstruction(Node * n) {
+    auto inst = insertInstruction(n->kind(), n->inputs(), moveFlags(n) , n->outputs());
+    instructions[inst].callback = getOperation(n, computes_on_variables);
+    return inst;
+  }
+  size_t insertInstruction(Symbol sym,
+                                 ArrayRef<Value*> inputs,
+                                 ArrayRef<uint8_t> move_flags,
+                                 ArrayRef<Value*> outputs) {
+    instructions.emplace_back();
+    auto & inst = instructions.back();
+    inst.debug_name = sym;
+    listBegin(inst.inputs.values);
+    listBegin(inst.inputs.free_flags);
```

Comment:
Can you have two lists "open" at the same time? Is it just because they are of different types? I would rather avoid such code, because it will be very easy to make mistakes